inst/include/dust/containers.hpp

```diff
@@ -95,36 +135,103 @@ class device_array {
   size_t size_;
 };
 
+// Specialisation of the above for void* memory needed by some cub functions
+// Construct once and use set_size() to modify
+// Still using malloc/free instead of new and delete, as void type problematic
+template <>
+class device_array<void> {
+public:
+  // Default constructor
+  device_array() : data_(nullptr), size_(0) {}
+  // Constructor to allocate empty memory
+  device_array(const size_t size) : size_(size) {
+    if (size_ > 0) {
+#ifdef __NVCC__
+      CUDA_CALL(cudaMalloc((void**)&data_, size_));
+#else
+      data_ = (void*) std::malloc(size_);
+      if (!data_) {
+        throw std::bad_alloc();
+      }
+#endif
+    }
+  }
+  ~device_array() {
+#ifdef __NVCC__
+    CUDA_CALL_NOTHROW(cudaFree(data_));
+#else
+    std::free(data_);
+#endif
+  }
+  void set_size(size_t size) {
+    size_ = size;
+#ifdef __NVCC__
+    CUDA_CALL(cudaFree(data_));
+    if (size_ > 0) {
+      CUDA_CALL(cudaMalloc((void**)&data_, size_));
+    } else {
+      data_ = nullptr;
+    }
+#else
+    std::free(data_);
+    if (size_ > 0) {
+      data_ = (void*) std::malloc(size_);
+      if (!data_) {
+        throw std::bad_alloc();
+      }
+    } else {
+      data_ = nullptr;
+    }
+#endif
+  }
+  void* data() {
+    return data_;
+  }
+  size_t size() const {
+    return size_;
+  }
+  size_t& size() {
+    return size_;
+  }
```

Comment:
yes, it returns a reference. I am confused who wrote this but it turns out that `cub::DeviceSelect::Flagged` may write to the size passed as the second arg. I've rewritten this to avoid this method which is likely dangerous.

https://github.com/mrc-ide/dustgpu/blob/cuda_shared/inst/include/dust/containers.cuh#L190-L195